### PR TITLE
Add Vietnamese input methods with shared VIME engine

### PR DESCRIPTION
**Pull request by plantaest**
Snapzy_2026-09-18_19-50-42_866

## Summary

This PR adds Vietnamese input methods to jQuery.IME:

- Telex
- Simple Telex
- VNI
- VIQR
- VIQR*
- traditional tone placement by default
- reformed tone-placement variants, labelled `đặt dấu kiểu mới`

The implementation uses one shared Vietnamese orthographic processor behind the input methods. Telex, VNI, VIQR, and VIQR* differ mainly in the keys they use; they share the same Vietnamese operations for tone placement, vowel diacritics, `d`/`đ`, `qu`/`gi`, repeated-key escape, tone reflow, and structural validation.

No jQuery.IME core API changes are required.

## Motivation

Vietnamese input in Wikimedia has been requested for a long time:

- Phabricator task: [T65465 - Add Vietnamese (vi) input method](https://phabricator.wikimedia.org/T65465)
- Related jQuery.IME discussion: [#60](https://github.com/wikimedia/jquery.ime/issues/60)
- Related per-input-method behavior discussion: [#342](https://github.com/wikimedia/jquery.ime/issues/342)

The original task already noted that Vietnamese users were relying on the local AVIM gadget, while some Wikimedia entry points such as special pages could benefit from integration through ULS/jQuery.IME. The discussion also pointed out that Vietnamese is not just a simple keyboard layout: users expect IME-like behavior, including delayed tone keys, tone relocation, literal fallback, and spelling-sensitive behavior.

Vietnamese Wikipedia has also discussed replacing the local AVIM gadget with ULS integration:

- Project start page: [Thay thế AVIM bằng ULS](https://vi.wikipedia.org/wiki/Thành_viên:Plantaest/Thay_thế_AVIM_bằng_ULS)
- Community consensus discussion: [Dự án Thay thế AVIM bằng ULS](https://vi.wikipedia.org/wiki/Wikipedia:Thảo_luận_cộng_đồng/Dự_án_Thay_thế_AVIM_bằng_ULS)
- Public testing/feedback discussion: [Bộ gõ tiếng Việt thay thế AVIM](https://vi.wikipedia.org/wiki/Wikipedia:Thảo_luận#Bộ_gõ_tiếng_Việt_thay_thế_AVIM)
- Project page: [VIME](https://vi.wikipedia.org/wiki/Thành_viên:Plantaest/VIME)
- Technical notes: [VIME technical specification](https://vi.wikipedia.org/wiki/Thành_viên:Plantaest/VIME/Đặc_tả_kỹ_thuật)

A compact demo was also published for community testing and feedback before this PR:

- Playground/demo: [https://taxon.toolforge.org/vime/](https://taxon.toolforge.org/vime/)

## Why a shared processor instead of only rule tables?

Many jQuery.IME input methods can be expressed well as direct pattern replacements. Vietnamese is harder to maintain that way because the same key can mean different operations depending on the current orthographic structure.

An earlier Vietnamese input-method attempt explored a regex/rule-table approach, but it became difficult to keep the behavior understandable and maintainable as more Vietnamese composition cases were added:

- 2021 project notes: [Xây dựng bộ gõ tiếng Việt trên nền jQuery.IME](https://vi.wikipedia.org/wiki/Thành_viên:Plantaest/Blog/Xây_dựng_bộ_gõ_tiếng_Việt_trên_nền_jQuery.IME)
- Related Phabricator comment: [T286863#7287345](https://phabricator.wikimedia.org/T286863#7287345)

Examples:

- A tone key represents a semantic tone, but the visible Unicode tone mark must be placed on the correct vowel.
- A later key may need to move an existing tone mark: `to1` can first produce `tó`, then `to1an` should become `toán`.
- Vowel-diacritic commands can be typed before or after more rime material.
- `qu` and `gi` need special handling.
- Telex, VNI, VIQR, and VIQR* express many of the same Vietnamese operations with different keys.
- Some Latin-looking input should remain literal when it is not a Vietnamese composition structure.

Encoding all of this independently as parallel regex tables would duplicate logic and edge cases across input methods. This PR instead separates:

1. input-method adapters, which translate keys into semantic commands; and
2. shared Vietnamese processing, which applies those commands to the current text.

At a high level:

```text
jQuery.IME key input
  -> Telex / Simple Telex / VNI / VIQR / VIQR* adapter
  -> shared Vietnamese command
  -> shared orthographic processor
  -> replacement text
```

This keeps method-specific mappings small while making behavior consistent across the supported Vietnamese input methods.

## Scope and compatibility

This change is scoped to Vietnamese input methods.

- It does not change the public jQuery.IME API.
- It does not change behavior for other input methods.
- It uses jQuery.IME's existing functional `patterns` path.
- It keeps all Vietnamese methods backed by one rule source: `rules/vi/vi.js`.
- It keeps Telex, Simple Telex, VNI, VIQR, and VIQR* as separate selectable input methods.
- It provides reformed tone-placement variants as separate input methods because jQuery.IME does not currently provide a per-input-method settings UI.

The processor is not a Vietnamese spell checker or dictionary. It uses finite orthographic recognition to decide whether a composition operation is structurally appropriate; it does not try to determine whether every possible syllable is a real word.

## Implementation notes

The Vietnamese implementation includes:

- semantic command decoding for Telex, Simple Telex, VNI, VIQR, and VIQR*;
- traditional and reformed tone placement;
- tone replacement and tone removal;
- vowel-diacritic composition;
- repeated-key escape for VNI and Telex-style commands;
- VIQR backslash escape;
- `qu` and `gi` handling;
- checked-syllable tone constraints;
- tone reflow when later input changes the resolved tone target;
- finite rime recognition for covered Vietnamese composition structures;
- pass-through behavior for unrecognized or structurally incompatible input;
- shifted-command handling through `patterns_shift` bridges where required by jQuery.IME.

The included `rules/vi/docs/` documents describe the requirements, architecture, algorithm, orthographic model, testing strategy, and terminology used by the implementation.

## Testing

The PR adds both pure VIME QUnit tests and jQuery.IME fixture coverage.

The tests cover:

- input-method registration and shared-source loading;
- Unicode parsing and rendering;
- parser and rime-recognizer behavior;
- tone placement and reformed tone placement;
- tone reflow;
- vowel-diacritic transformation;
- post-transform validation;
- adapter behavior;
- Telex, Simple Telex, VNI, VIQR, and VIQR* typing sequences;
- shifted command keys;
- repeated-key and VIQR escape behavior;
- representative regression cases from manual testing and community feedback.

Focused Vietnamese tests and the relevant jQuery.IME fixture tests pass locally.

## Review note

Vietnamese input behavior has more context-sensitive state than most direct keyboard-layout rules. The implementation is therefore organized as small shared orthographic operations rather than a flat rule table. Reviewing the tests together with `rules/vi/vi.js` may make the intended behavior easier to follow.

I am happy to make small follow-up fixes or adjustments during review if reviewers or Vietnamese users find edge cases that should be covered before merge.

## Related

- Phabricator: [T65465](https://phabricator.wikimedia.org/T65465)
- GitHub: [#60](https://github.com/wikimedia/jquery.ime/issues/60)
- GitHub: [#342](https://github.com/wikimedia/jquery.ime/issues/342)

### CODE REVIEW

File: rules/vi/docs/status.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+supported scope.
+
+## Known limitations
+
+The recognizer validates written structure, not lexical existence. A
+Vietnamese-shaped nonce syllable can still be accepted if its structure is in
+the covered model.
+
+Rare, dialectal, minority-language, historical, and specialized spellings may
+need new structural examples and tests before the recognizer should accept them.
```
Comment (1ec5) [inline rules/vi/docs/status.md]: I put together [a battery of tests](https://github.com/1ec5/avim/blob/18572ba2ded98cb5127b5f7ffda6b2b5e5ca3dca/tests/test.js) for AVIM for Firefox based on bug reports filed against various Vietnamese IMEs over the years. It covers some edge cases like [_giặt gỵa_](https://en.wiktionary.org/wiki/gi%E1%BA%B7t_g%E1%BB%B5a#Vietnamese) that would only be nice to have, as well as some more common cases in Telex that might be more important to handle in this initial PR.

File: rules/vi/docs/status.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+Some Telex ambiguity remains inherent in a rendered-text-first jQuery.IME
+adapter. Full Telex uses a narrow raw-key context only for standalone `w`
+quick-key escape. Other Vietnamese behavior still prefers documented structural
+behavior and repeated-key escape rather than persistent raw-key history or a
+user-visible spell-check option.
+
+## Possible follow-up topics
+
+Future work may include:
```
Comment (1ec5) [inline rules/vi/docs/status.md:131]: Another possible future direction might be to integrate a machine learning or language model for adding diacritics to unaccented text. (Or candidate lists, similar to CJK IMEs.) To date, few input methods have implemented this feature specifically for Vietnamese. Only mobile keyboards with their built-in spelling dictionaries and predictive text engines support something similar, but somewhat unpredictably.

File: test/jquery.ime.test.fixtures.js
Comment (1ec5) [inline test/jquery.ime.test.fixtures.js:1]: At a glance, this test suite is quite substantial, but I think a more comprehensive [dictionary-based integration test](https://phabricator.wikimedia.org/T65465#7287352) would give us more confidence that switching a site from AVIM to VIME wouldn’t introduce any regressions. Would something like that be feasible within this repository, or would it have to be a separate project?

File: rules/vi/docs/requirements.md [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+avoids rendering nonstandard checked-tone forms while keeping the user's literal
+command available.
+
+## Unicode behavior
+
+The implementation MUST produce valid Unicode Vietnamese text.
+
+Rendered output SHOULD use NFC unless a documented jQuery.IME or browser constraint requires another form.
+
+The engine SHOULD correctly interpret canonically equivalent input where practical. It must not assume that every visible Vietnamese character is one JavaScript code unit.
```
Comment (1ec5) [inline rules/vi/docs/requirements.md]: Would it be feasible to automatically substitute the Icelandic ð, which looks similar in some fonts? 1ec5/avim#34

File: rules/vi/docs/requirements.md [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+
+## Non-Vietnamese text
+
+The first priority is correct Vietnamese composition. Conservative protection
+against all foreign words, code identifiers, or technical text is not required
+for the current engine boundary.
+
+The architecture SHOULD allow stricter structural validation later, but VIME
+MUST NOT add a Vietnamese dictionary dependency merely to avoid accidental
+transformations.
```
Comment (1ec5) [inline rules/vi/docs/requirements.md]: Does jquery.ime support binding a keystroke to terminating a composition and starting a new one? 1ec5/avim#14 binds Shift+Space to a syllable break.[^dom]

The user would be able to type loanwords more easily by pressing the keystroke after each syllable. For example: _kilômét_, _véctơ_, _México_, _Phnôm Pênh_, _Oasinhtơn_, _Grêgôriô_. It would be much more convenient than inserting temporary hyphens or spaces. This respelling style has been abandoned in some fields in Vietnam (such as chemistry), but it still persists in other fields (such as in geography and Catholicism) as well as abroad.

[^dom]: It’s a bit of a hack. The engine doesn’t track keystroke history. Instead, the UI inserts an invisible element into the text input field’s shadow DOM that prevents the engine from seeing or manipulating any preceding text. The browser drops the element when serializing the field’s value as plain text.

File: rules/vi/docs/requirements.md [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+```text
+a'        -> á
+a`        -> à
+a?        -> ả
+a~        -> ã
+a.        -> ạ
+a^        -> â
+a(        -> ă
+o+        -> ơ
+dd        -> đ
```
Comment (1ec5) [inline rules/vi/docs/requirements.md:290]: Some IMEs automatically turn `đ`<kbd>_</kbd> into ₫ or `d`<kbd>d</kbd> into ₫ following a number, making it easier to enter properly typeset prices in Vietnamese đồng.

/ref 1ec5/avim#92

Comment (hideonrosie) [reply rules/vi/docs/requirements.md:290]: @1ec5 It seems that only AVIM has this feature. OpenKey, UniKey, and EVKey have never had it.

Comment (1ec5) [reply rules/vi/docs/requirements.md:292]: The heuristic for doing the conversion after a number is unique to AVIM, but some virtual keyboards have affordances for the ₫ symbol. At least when I last updated [this feature comparison](https://avim.1ec5.org/compare.html) some years ago, the virtual keyboard built into Google Docs (also available as a Chrome extension) had a key binding for it in TCVN mode.

Comment (hideonrosie) [reply rules/vi/docs/requirements.md:292]: @1ec5 I don't think the existence of such a shortcut is, by itself, a strong reason to treat it as a requirement. The Google Docs virtual keyboard supporting đ in TCVN mode is a fairly specific use case, and I'm not convinced that it represents a common expectation among Vietnamese users today.

Most modern Vietnamese text input is Unicode-based, while TCVN-based input is considerably more niche. There is also a difference between a feature being available in a particular virtual keyboard and being a generally expected Vietnamese input convention. I would therefore hesitate to encode this kind of niche behavior in the general requirements.

I'd be inclined to keep the automatic d → đ conversion after a number as an AVIM-specific heuristic/convenience, unless we have evidence that this behavior is sufficiently widespread to warrant standardizing it.

P/S: I just tested extension myself, but I couldn't reproduce the behavior. Is this still supported?

Comment (1ec5) [reply rules/vi/docs/requirements.md:292]: That’s fair, I didn’t fully grasp the difference between the requirements document and the intended scope of the PR. It’s a niche enough behavior that we could certainly treat it as a separate enhancement request.

I don’t think Google has developed their IME extension in many years, apart from compatibility updates. The feature continues to be built into Google Docs though.

File: rules/vi/docs/status.md (hunk 1 above)
Comment (plantaest) [reply rules/vi/docs/status.md]: Hi @1ec5, it's great to see you again after such a long time. I hope you've been well.

Thanks for the AVIM `test.js`. I used it as a regression checklist rather than importing it wholesale, because AVIM has runtime preferences such as `ckSpell`, `informal`, and `oldAccent`, while this jQuery.IME implementation currently has one fixed structural model and no user-visible spell-check toggle.

I pushed a follow-up commit (ee62093) that does a few things:

* Fixes default Telex `quw`. Previously VIME treated `qu` like an onset-only prefix and allowed quick-`w`, producing `quư`. Now `qu` is treated as a special onset for this path, so `quw` stays literal while cases such as `tw -> tư` still work.
* Adds fixture coverage inspired by the AVIM tests for common Telex/VIQR behavior, including `kius`, `kis`, `tyr`, `khuyur`, `to'an`, `truong+`, truơng&#96;, `queo.`, `queue.`, zi&#96;, and `qi'`.
* Documents the current boundary around contextual spelling validation.

There are a few intentional differences from AVIM:

* VIME does not currently enforce deeper contextual spelling pairs such as `c/k`. So cases like `cys -> cý`, `coaf -> còa`, `coef -> còe`, and `cuyf -> cùy` are accepted for now because their rimes are structurally covered. This matches the current goal of VIME as a composition engine, not a full spell checker. OpenKey behaves similarly even with spell checking enabled.
* For `gya.`, AVIM distinguishes behavior depending on `ckSpell`; VIME has no such mode, so it accepts the rare spelling and composes `gỵa`.
* For `khuya'`, VIME intentionally supports the rare `uya` rime and places the tone on `y`, producing `khuýa`.

I did not import AVIM-specific behavior such as `₫`, hyphen-based undo, dotless `ı`, or preference-dependent spell-check behavior, since those would need separate design decisions and probably should not block this initial PR.

I'll continue reviewing the other comments. Please give me some time, and thank you very much.

File: test/jquery.ime.test.fixtures.js
Comment (plantaest) [reply test/jquery.ime.test.fixtures.js:1]: Yes, I think this is feasible and would be very valuable.

My preference would be to keep the corpus test harness in this repository, but not necessarily vendor the corpus itself or run it as part of the default QUnit suite. For example, an optional runner could accept an external dictionary path:

    node rules/vi/vi.test.corpus.js path/to/vi.dic

The runner could generate representative input sequences for each supported Vietnamese input method, feed them through the same VIME processor used by the input methods, and report entries that fail to round-trip to the expected output.

I would treat this primarily as migration/integration validation rather than ordinary unit coverage. The normal suite should continue to contain focused deterministic regression cases, while the corpus runner could provide much broader confidence before switching viwiki from AVIM to VIME.

I would also prefer to keep the corpus external initially, both to avoid adding a large data file and to avoid making assumptions about licensing or attribution before we decide on a specific corpus.

I would be happy to build this as a follow-up. Any useful failures found by the corpus run could then be reduced to focused cases and added to the regular VIME regression fixtures.

File: rules/vi/docs/status.md (hunk 2 above)
Comment (plantaest) [reply rules/vi/docs/status.md:133]: I have thought about this before, although separately from VIME. I once came across [v7](https://github.com/ducngg/v7), which appears to experiment with an AI/model-assisted Vietnamese input method.

For VIME, though, I would prefer to keep the scope deterministic and rule-based, within the existing jQuery.IME architecture. Predictive diacritic restoration or candidate generation would require a rather different architecture and UX, so I would consider that more suitable for a separate project than a future VIME feature.

Still, it is an interesting direction.

File: rules/vi/docs/requirements.md (hunk 3 above)
Comment (plantaest) [reply rules/vi/docs/requirements.md]: I would prefer not to include this in the initial VIME implementation.

`ð`/`Ð` are legitimate characters in other languages rather than malformed Vietnamese text, and Wikipedia is a fairly multilingual editing environment. Automatically normalizing them to `đ`/`Đ` could therefore introduce surprising changes in names, quotations, code, or other non-Vietnamese text.

This also fits VIME's current conservative approach: if the input does not clearly look like a Vietnamese composition, it should generally be left unchanged.

I can see value in legacy-text normalization, especially for text copied from older Vietnamese encodings or sources, but I think that would be better treated as a separate, explicitly scoped feature rather than part of the core typing rules.

File: rules/vi/docs/requirements.md (hunk 5 above)
Comment (plantaest) [reply rules/vi/docs/requirements.md:292]: Thanks. I would still prefer **not** to implement this in VIME.

I see conversions such as turning `đ_` into `₫`, or interpreting `dd` as `₫` after a number, as context-sensitive typographic shortcuts rather than Vietnamese orthographic input rules. They are useful conveniences, but conceptually closer to text expansion or symbol-entry features than to Telex/VNI/VIQR composition itself.

VIME is intentionally scoped to deterministic Vietnamese character and diacritic composition, so I would rather keep this kind of currency-symbol heuristic outside the core engine.

I agree that it could be considered separately as an enhancement if there is enough demand, but I would leave it out of this PR and the current VIME requirements.

File: rules/vi/docs/requirements.md (hunk 4 above)
Comment (plantaest) [reply rules/vi/docs/requirements.md]: Thanks for pointing this out. I added initial support for this in 16bbf78.

Within the current jQuery.IME extension points, VIME can bind Shift+Space through `patterns_shift`, so the implementation now treats Shift+Space as an invisible composition boundary for the Vietnamese input methods. Regular Space still inserts a visible space.

The implementation is a little different from AVIM’s DOM-marker approach. VIME does not insert an invisible node or sentinel into the editable content. Instead, the Vietnamese adapter records a bounded rendered-text state at the boundary. On subsequent keystrokes, if the current input window still matches that state, the adapter scopes the transformation to only the suffix after the boundary and reattaches the frozen prefix.

For example, this lets users type cases like:

```text
VNI:   ki Shift+Space lo6 Shift+Space met1 -> kilômét
Telex: ki Shift+Space loo Shift+Space mets -> kilômét
VNI:   ve1c Shift+Space to7 -> véctơ
Telex: a Shift+Space s k -> ask
```

This should cover the main use case you described: composing respelled loanwords as several logical syllables without inserting temporary visible spaces or hyphens.

There are some constraints worth noting. jQuery.IME’s current `patterns_shift` bridge is array-based, so this is not a full core-level composition-boundary primitive. The boundary is deliberately self-invalidating: if the surrounding rendered text no longer matches the stored bounded state, VIME clears the boundary and processes the full input normally. This keeps the feature local to the Vietnamese adapter and avoids changing jQuery.IME core.

I also added unit coverage for suffix scoping and integration fixtures for Shift+Space typing sequences, and documented the behavior in the VIME docs.

**Comment by plantaest** [comment]
Hi @1ec5, I added an optional corpus audit script in 754c9f7, modeled after the same broad idea as AVIM’s corpus test: generate canonical keystrokes from a Unicode word list, then type those keys through the actual registered jQuery.IME input methods.

I ran it locally against both hunspell-vi dictionaries:

- https://github.com/1ec5/hunspell-vi/blob/main/dictionaries/vi-DauCu.dic
- https://github.com/1ec5/hunspell-vi/blob/main/dictionaries/vi-DauMoi.dic

The current results are:

```text
traditional telex:        6630/6642 passed, 12 failed
traditional telex-simple: 6632/6642 passed, 10 failed
traditional vni:          6637/6642 passed, 5 failed
traditional viqr:         6637/6642 passed, 5 failed
traditional viqr-star:    6637/6642 passed, 5 failed

reformed telex:           6630/6642 passed, 12 failed
reformed telex-simple:    6632/6642 passed, 10 failed
reformed vni:             6637/6642 passed, 5 failed
reformed viqr:            6637/6642 passed, 5 failed
reformed viqr-star:       6637/6642 passed, 5 failed
```

The first corpus run did uncover a real bug: open `quy` was being analyzed in a way that placed tone marks on `u` (`qúy`) instead of `y` (`quý`). That is fixed in the same commit, with regression coverage for VNI, Telex, and VIQR.

I also triaged the remaining failures. They do not currently look like ordinary Vietnamese composition bugs:

- Telex failures such as `ASCII`, `GIF`, `URL`, `VISCII`, `TW`, `web`, and `basoi` are literal-text conflicts with Telex command keys. They can be typed with repeated-key escape, for example `ASSCII`, `GIFF`, `URRL`, `VISSCII`, `TWW`, `wweb`, and `bassoi`. `web` also works directly in Simple Telex.
- `Lắk` and `Đắk` fail only in the corpus-generated canonical order (`Lak81`, `Dak981`), where the command comes after final `k`. They are still typeable by applying the marks before `k`, for example `La81k` and `D9a81k`.
- `palăng` is a loanword-like multi-part case. It is typeable with the new invisible composition boundary, for example `pa` + Shift+Space + `lang8`.
- `HĐND` is an acronym with `Đ` inside it. It is also typeable with composition boundaries, for example `H` + Shift+Space + `D9` + Shift+Space + `ND`.
- `Nguơn` conflicts with the normal `ươn` promotion path. It can be forced with a boundary, for example `Nguo7` + Shift+Space + `n`.

So I think this gives us a useful confidence check for AVIM replacement: the corpus audit caught one real issue, and after fixing it, the remaining mismatches are explainable as ambiguity, canonical-order limitations, acronym/loanword boundaries, or spellings outside the default structural model. The script is intentionally optional because the corpora are external, but it should be useful for future regression checks and for comparing behavior during deployment.
